README.txt
The Produce & Publish Client Connector for Plone
================================================

Documentation
-------------

See http://packages.python.org/zopyx.smartprintng.plone

Source code
-----------
See https://github.com/zopyx/zopyx.smartprintng.plone

Bugtracker
----------
See  https://github.com/zopyx/zopyx.smartprintng.plone/issues

Licence
-------
Published under the GNU Public Licence Version 2 (GPL 2)

Author
------
| ZOPYX Limited
| Charlottenstr. 37/1
| D-72070 Tuebingen, Germany
| info@zopyx.com
| www.zopyx.com




docs/HISTORY.rst
Changelog
=========

2.1.9 (2011-11-11)
------------------
- fixed bug in makeImagesLocal() transformation
  where the document root has not been used properly
  for finding images by traversal 

2.1.8 (2011-11-11)
------------------
- support for new ``SMARTPRINTNG_LOCAL_CONVERSION`` environment
  variable

2.1.7 (2011-11-08)
------------------
- removed some baggage in order to make distro smaller

2.1.6 (2011-11-07)
------------------
- minor fixes in handling of generated files for download 

2.1.5 (2011-11-07)
------------------
- first public (open-source) release

2.1.4 (2011-10-25)
------------------
- fixed unicode/utf-8 issue in makeImagesLocal transformation

2.1.3 (2011-10-14)
------------------
- added fixAmpersand transformation

2.1.2 (2011-10-10)
------------------
- transformations for dealing with converted footnotes from Word

2.1.1 (2011-10-08)
------------------
- compatibility with Dexterity

2.1.0 (2011-09-22)
------------------
- final 2.1.0 release

2.0.9 (2011-09-20)
------------------
- fixed bug in xpath_query() (using relative query)

2.0.8 (2011-09-11)
------------------
- more cleanup

2.0.7 (2011-09-10)
------------------
- some ZCML fixes in order to avoid Plone 4.x startup failures under
  some conditions
- restored compatibility with Plone 3.3.X

2.0.6 (2011-09-08)
------------------
- image exporter did not deal proper with duplicate image ids
- minor fixes

2.0.5 (2011-09-02)
------------------
- new lxml backed transformation pipeline 
- more tests

2.0.4 (2011-08-26)
------------------
- logging resource registration using INFO severity
- new lxml dependency

2.0.3 (2011/08/15)
------------------
- catching HTTPError in image resolver
- fixed another BeautifulSoup misbehaviour in fixHeadingAfterOfficeImport()

2.0.2 (2011-08-02)
------------------
- minor fix

2.0.1 (2011-08-02)
------------------
- integration with new zip client version (Proxy2 implementation)

2.0.0 (2011-07-25)
---------------------
* final release

2.0.0rc2 (2011-07-04)
---------------------
* fix in logger call in folder.py

2.0.0rc1 (2011-07-01)
---------------------
* don't extend images an authoring project
* remove class attributes from headings after office import
* added ignoreHeadingsForStructure transformation

2.0.0b2 (2011-06-16)
--------------------
* minor fixes related to office data import

2.0.0b1 (2011-05-24)
--------------------
* fixes related to office format input

2.0.0a3 (2011-05-17)
--------------------
* added some workaround for image resolver in order to deal with images
  referenced through a fully specified URL with a redirection included
  (TQM issue)

2.0.0a2 (2011-05-14)
--------------------
* minor fix in safe_get()

2.0.0a1 (2011-05-10)
--------------------
* simplification and refacoring

0.7.0 (2011-02-11)
-------------------
* updated for use with zopyx.authoring 1.5.X
* added GenericDownloadView aka '@@ppConvert'
* exported images now contain a proper extension (fixes issues
  with the XFC converter depending on extension for determining
  the image format)

0.6.24 (2010-12-09)
-------------------
* added addDocumentLinks() transformation
* including content ids of aggregated content

0.6.23 (2010-09-10)
-------------------
* addImageCaptionsInHTML(): honour excludeFromImageEnumeration

0.6.22 (2010-09-09)
-------------------
* fixed improper stripping of image names using an image scale
  (causing issues in the consolidated HTML view of the authoring
  environment)

0.6.21 (2010-08-09)
-------------------
* added support '++resource++' image references (Patrick Gerken)
* added support for FSImage (Patrick Gerken)

0.6.20 (2010-08-05)
-------------------
* added 'removeComments' transformation
* added 'makeImageSrcLocal' transformation

0.6.19 (2010-07-13)
-------------------
* fixed race condition in makeImagesLocal()

0.6.18 (2010-06-14)
-------------------
* images got a new PDF conversion option "Exclude from image enumeration"

0.6.17 (2010-06-12)
-------------------
* inserting H1 title for consolidated HTML
* added extra class to folder title for consolidated HTML 

0.6.16 (2010-05-29)
-------------------
* inserting space for found anchors

0.6.15 (2010-04-15)
-------------------
* minor fix in image handling

0.6.14 (2010-04-14)
-------------------
* minor tweaks for image caption markup

0.6.13 (2010-03-26)
-------------------
* support for span.footnoteText

0.6.12 (2010-03-21)
-------------------
* support for image urls 'resolveuid/<uid>'
* minor fixes and tweaking in image handling (caption generation)

0.6.11 (2010-03-10)
-------------------
* added document extender
* document option for suppressing the title in PDF
* image caption support
* changed default transformations (to makeImagesLocal only)
* removed TOC from default PDF template

0.6.10 (2010-03-03)
-------------------
* support for request/transformations parameter
* various fixes

0.6.9 (2010-02-22)
------------------
* added <em>[[text:footnote-text]]</em> support for generating footnotes
* various changes related to zopyx.authoring integration

0.6.8 (2010-02-03)
------------------

* Folder aggregation now works with all folderish objects providing IATFolder


0.6.7 (2009-11-30)
------------------

* makeImagesLocal: better dealing with virtual hosting

0.6.6 (2009-11-15)
------------------

* fixed CSS issue with TOC markup

0.6.5 (2009-11-12)
------------------

* always use images in their original resolution 
* optional content information with link to the edit mode
  of the aggregated document (you must change the visibility
  of the .content-info class through CSS)
* a request parameter 'show-debug-info' will enable the
  additional content-info view
* better error handling
* better logging
* tweaked markup of generated TOC


0.6.3 (2009-10-27)
------------------

* refactored language handling
* refactored PDF view in order to provide a low-level view 
  returning a reference to the generated PDF file instead
  providing it for HTTP download



zopyx/smartprintng/plone/browser/pdf.py
################################################################
# zopyx.smartprintng.plone
# (C) 2011,  ZOPYX Limited & Co. KG, D-72070 Tuebingen, Germany
################################################################

import os
import glob
import codecs
import shutil
import tempfile
import urllib2
import zipfile

from compatible import InitializeClass
from Products.Five.browser import BrowserView
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
from Products.ATContentTypes.interface.folder import IATFolder
from ZPublisher.Iterators import filestream_iterator
from zope.app.pagetemplate.viewpagetemplatefile import ViewPageTemplateFile as ViewPageTemplateFile2

from zopyx.smartprintng.plone.logger import LOG
from zopyx.smartprintng.plone.resources import resources_registry
from zopyx.smartprintng.plone import Transformer

import splitter
from util import getLanguageForObject

cwd = os.path.dirname(os.path.abspath(__file__))

# server host/port of the SmartPrintNG server
URL = os.environ.get('SMARTPRINTNG_SERVER', 'http://localhost:6543')
LOCAL_CONVERSION = 'SMARTPRINTNG_LOCAL_CONVERSION' in os.environ
ZIP_OUTPUT = 'SMARTPRINTNG_ZIP_OUTPUT' in os.environ

class ProducePublishView(BrowserView):
    """ Produce & Publish view (using Produce & Publish server) """

    template = ViewPageTemplateFile('../resources_default/pdf_template_standalone.pt')

    # default transformations used for the default PDF view.
    # 'transformations' can be overriden within a derived ProducePublishView.
    # If you don't need any transformation -> redefine 'transformations'
    # as empty list or tuple

    transformations = (
        'makeImagesLocal',
#        'removeEmptyElements',
#        'removeInternalLinks',
#        'annotateInternalLinks',
#        'cleanupTables',
        'convertFootnotes',
        'removeCrapFromHeadings',
        'fixHierarchies',
        'addTableOfContents',
        )

    def copyResources(self, resources_dir, destdir):
        """ Copy over resources for a global or local resources directory into the 
            destination directory.
        """
        if os.path.exists(resources_dir):
            for name in os.listdir(resources_dir):
                fullname = os.path.join(resources_dir, name)
                if os.path.isfile(fullname):
                    shutil.copy(fullname, destdir)

    def transformHtml(self, html, destdir, transformations=None):
        """ Perform post-rendering HTML transformations """

        if transformations is None:
            transformations = self.transformations

        # the request can override transformations as well
        if self.request.has_key('transformations'):
            t_from_request = self.request['transformations']
            if isinstance(t_from_request, basestring):
                transformations = t_from_request and t_from_request.split(',') or []
            else:
                transformations = t_from_request

        T = Transformer(transformations, context=self.context, destdir=destdir)
        return T(html)

    def __call__(self, *args, **kw):

        try:
            return self.__call2__(*args, **kw)
        except:
            LOG.error('Conversion failed', exc_info=True)
            raise


    def __call2__(self, *args, **kw):
        """ URL parameters:
            'language' -  'de', 'en'....used to override the language of the
                          document
            'converter' - default to on the converters registered with
                          zopyx.convert2 (default: pdf-prince)
            'resource' - the name of a registered resource (directory)
            'template' - the name of a custom template name within the choosen
                         'resource' 
        """

        # Output directory
        tmpdir_prefix = os.path.join(tempfile.gettempdir(), 'produce-and-publish')
        if not os.path.exists(tmpdir_prefix):
            os.makedirs(tmpdir_prefix)
        destdir = tempfile.mkdtemp(dir=tmpdir_prefix, prefix=self.context.getId() + '-')

        # debug/logging
        params = kw.copy()
        params.update(self.request.form)
        LOG.info('new job (%s, %s) - outdir: %s' % (args, params, destdir))

        # get hold of the language (hyphenation support)
        language = getLanguageForObject(self.context)
        if params.get('language'):
            language = params.get('language')

        # Check for CSS injection
        custom_css = None
        custom_stylesheet = params.get('custom_stylesheet')
        if custom_stylesheet:
            custom_css = str(self.context.restrictedTraverse(custom_stylesheet, None))
            if custom_css is None:
                raise ValueError('Could not access custom CSS at %s' % custom_stylesheet)

        # check for resource parameter
        resource = params.get('resource')
        if resource:
            resources_directory = resources_registry.get(resource)
            if not resources_directory:
                raise ValueError('No resource "%s" configured' % resource)
            if not os.path.exists(resources_directory):
                raise ValueError('Resource directory for resource "%s" does not exist' % resource)
            self.copyResources(resources_directory, destdir)

            # look up custom template in resources directory
            template_name = params.get('template', 'pdf_template')
            if not template_name.endswith('.pt'):
                template_name += '.pt'
            template_filename = os.path.join(resources_directory, template_name)
            if not os.path.exists(template_filename):
                raise IOError('No template found (%s)' % template_filename)
            template = ViewPageTemplateFile2(template_filename)

        else:
            template = self.template

        # call the dedicated @@asHTML on the top-level node. For a leaf document
        # this will return either a HTML fragment for a single document or @@asHTML
        # might be defined as an aggregator for a bunch of documents (e.g. if the
        # top-level is a folderish object

        html_view = self.context.restrictedTraverse('@@asHTML', None)
        if not html_view:
            raise RuntimeError('Object at does not provide @@asHTML view (%s, %s)' % 
                               (self.context.absolute_url(1), obj.portal_type))
        html_fragment = html_view()

        # arbitrary application data
        data = params.get('data', None)

        # Now render the complete HTML document
        html = template(self,
                        language=language,
                        request=self.request,
                        body=html_fragment,
                        custom_css=custom_css,
                        data=data,
                        )

        # and apply transformations
        html = self.transformHtml(html, destdir)

        # hack to replace '&' with '&amp;'
        html = html.replace('& ', '&amp; ')

        # and store it in a dedicated working directory
        dest_filename = os.path.join(destdir, 'index.html')
        fp = codecs.open(dest_filename, 'wb', encoding='utf-8')
        fp.write(html)
        fp.close()  

        # split HTML document into parts and store them on the filesystem
        # (making only sense for folderish content)
        if IATFolder.providedBy(self.context) and not 'no-split' in params:
            splitter.split_html(dest_filename, destdir)

        # copy over global styles etc.
        resources_dir = os.path.join(cwd, 'resources')
        self.copyResources(resources_dir, destdir)

        # copy over language dependent hyphenation data
        if language:
            hyphen_file = os.path.join(resources_dir, 'hyphenation', language + '.hyp')
            if os.path.exists(hyphen_file):
                shutil.copy(hyphen_file, destdir)

            hyphen_css_file = os.path.join(resources_dir, 'languages', language + '.css')
            if os.path.exists(hyphen_css_file):
                shutil.copy(hyphen_css_file, destdir)

        # now copy over resources (of a derived view)
        self.copyResources(getattr(self, 'local_resources', ''), destdir)
        if ZIP_OUTPUT or 'zip_output' in params:
            archivename = tempfile.mktemp(suffix='.zip')
            fp = zipfile.ZipFile(archivename, "w", zipfile.ZIP_DEFLATED) 
            for root, dirs, files in os.walk(destdir):
                #NOTE: ignore empty directories
                for fn in files:
                    absfn = os.path.join(root, fn)
                    zfn = absfn[len(destdir)+len(os.sep):] #XXX: relative path
                    fp.write(absfn, zfn)
            fp.close()
            LOG.info('ZIP file written to %s' % archivename)

        if 'no_conversion' in params:
            return destdir
        
        if LOCAL_CONVERSION:
            from zopyx.convert2 import Converter
            c = Converter(dest_filename)
            result = c(params.get('converter', 'pdf-pisa'))
            if result['status'] != 0:
                raise RuntimeError('Error during PDF conversion (%r)' % result)
            pdf_file = result['output_filename']
            LOG.info('Output file: %s' % pdf_file)
            return pdf_file
        else:
            # Produce & Publish server integration
            from zopyx.smartprintng.client.zip_client import Proxy2
            proxy = Proxy2(URL)
            result = proxy.convertZIP2(destdir, self.request.get('converter', 'pdf-prince'))
            LOG.info('Output file: %s' % result['output_filename'])
            return result['output_filename']

InitializeClass(ProducePublishView)


class PDFDownloadView(ProducePublishView):

    def __call__(self, *args, **kw):

        if not 'resource' in kw:
            kw['resource'] = 'pp-default'
        if not 'template' in kw:
            kw['template'] = 'pdf_template_standalone'
        kw['no-split'] = True

        output_file = super(PDFDownloadView, self).__call__(*args, **kw)
        mimetype = os.path.splitext(os.path.basename(output_file))[1]
        R = self.request.response
        R.setHeader('content-type', 'application/%s' % mimetype)
        R.setHeader('content-disposition', 'attachment; filename="%s.%s"' % (self.context.getId(), mimetype))
        R.setHeader('pragma', 'no-cache')
        R.setHeader('cache-control', 'no-cache')
        R.setHeader('Expires', 'Fri, 30 Oct 1998 14:19:41 GMT')
        R.setHeader('content-length', os.path.getsize(output_file))
        return filestream_iterator(output_file, 'rb').read()

InitializeClass(PDFDownloadView)


class GenericDownloadView(ProducePublishView):

    def __call__(self, *args, **kw):

        output_file = super(GenericDownloadView, self).__call__(*args, **kw)
        mimetype = os.path.splitext(os.path.basename(output_file))[1]
        # return output file over HTTP
        R = self.request.response
        R.setHeader('content-type', 'application/%s' % mimetype)
        R.setHeader('content-disposition', 'attachment; filename="%s.%s"' % (self.context.getId(), mimetype))
        R.setHeader('content-length', os.path.getsize(output_file))
        R.setHeader('pragma', 'no-cache')
        R.setHeader('cache-control', 'no-cache')
        R.setHeader('Expires', 'Fri, 30 Oct 1998 14:19:41 GMT')
        return filestream_iterator(output_file, 'rb')

InitializeClass(GenericDownloadView)
